File: parser/docCheck.js
/**
 * docCheck.js
 * Audits a parsed NodeGraph for documentation health.
 *
 * Checks:
 *   1. Every directory that contains source files has a README.md
 *   2. Every README.md was modified more recently than the files in its directory
 *   3. Flags files with no exports and no definitions as "orphan" candidates
 *
 * Returns a DocAudit object that gets merged into the NodeGraph meta.
 * The frontend uses this to render warning badges on nodes/folders.
 */

import { readdirSync, statSync } from 'fs';
import { join, dirname, basename } from 'path';

/**
 * @param {string} rootDir
 * @param {import('./projectParser.js').FileNode[]} nodes
 * @returns {DocAudit}
 */
export function auditDocs(rootDir, nodes) {
  const warnings = [];

  // Build a map: dirPath -> { readmeMtime, files: [{path, mtime}] }
  const dirMap = new Map();

  for (const node of nodes) {
    if (node.isExternal) continue;
    const dir = dirname(join(rootDir, node.path));
    const rel = dir.replace(rootDir, '').replace(/^[/\\]/, '') || '.';

    if (!dirMap.has(rel)) {
      dirMap.set(rel, { readmeMtime: null, readmePath: null, files: [] });
    }

    const entry = dirMap.get(rel);

    if (node.ext === '.md' && basename(node.path).toLowerCase() === 'readme.md') {
      entry.readmeMtime = node.modified ? new Date(node.modified).getTime() : null;
      entry.readmePath = node.path;
    } else {
      // Only count parseable source files, not docs themselves
      const ignoredExts = new Set(['.md', '.mdx', '.txt', '.lock', '.log']);
      if (!ignoredExts.has(node.ext)) {
        entry.files.push({
          path: node.path,
          mtime: node.modified ? new Date(node.modified).getTime() : null,
          id: node.id,
        });
      }
    }
  }

  // ── Check 1: Missing README ──────────────────────────────────────────────
  for (const [dir, entry] of dirMap) {
    if (entry.files.length === 0) continue; // skip dirs with only docs

    if (!entry.readmeMtime) {
      warnings.push({
        type: 'missing-readme',
        severity: 'warning',
        dir,
        message: `No README.md in ${dir || 'root'}`,
        affectedFiles: entry.files.map(f => f.id),
      });
      continue;
    }

    // ── Check 2: Stale README ──────────────────────────────────────────────
    const staleFiles = entry.files.filter(f =>
      f.mtime !== null && entry.readmeMtime !== null && f.mtime > entry.readmeMtime
    );

    if (staleFiles.length > 0) {
      warnings.push({
        type: 'stale-readme',
        severity: 'info',
        dir,
        readmePath: entry.readmePath,
        message: `README.md in ${dir || 'root'} is older than ${staleFiles.length} file(s)`,
        affectedFiles: staleFiles.map(f => f.id),
        staleSince: new Date(Math.max(...staleFiles.map(f => f.mtime))).toISOString(),
      });
    }
  }

  // ── Check 3: Undocumented nodes (no exports, no definitions, not a config/doc) ──
  const undocumentedCategories = new Set(['javascript', 'typescript', 'react', 'python', 'shell']);
  for (const node of nodes) {
    if (node.isExternal) continue;
    if (!undocumentedCategories.has(node.category)) continue;
    if ((node.definitions?.length || 0) === 0 && (node.exports?.length || 0) === 0) {
      // Could be a pure side-effect file or an entrypoint — only warn if it also has no imports
      if ((node.imports?.length || 0) === 0) {
        warnings.push({
          type: 'empty-module',
          severity: 'info',
          nodeId: node.id,
          message: `${node.path} has no imports, exports, or definitions`,
          affectedFiles: [node.id],
        });
      }
    }
  }

  const missingReadmes = warnings.filter(w => w.type === 'missing-readme').length;
  const staleReadmes = warnings.filter(w => w.type === 'stale-readme').length;
  const emptyModules = warnings.filter(w => w.type === 'empty-module').length;

  return {
    healthy: warnings.length === 0,
    summary: {
      missingReadmes,
      staleReadmes,
      emptyModules,
      total: warnings.length,
    },
    warnings,
  };
}

File: parser/doc_check.py
# SPDX-License-Identifier: GPL-3.0-or-later
# Copyright (C) 2026 N0V4-N3XU5

"""
parser/doc_check.py
====================
Audits a parsed project for documentation health and emits warnings
that the node editor renders as badges on files and directories.

Checks performed
-----------------
1. MISSING README  — a directory that contains source files but has no README.md
2. STALE README    — a README.md that is older than one or more files in its dir
3. EMPTY MODULE    — a source file with no imports, exports, or definitions
                     (likely a forgotten stub or accidentally blank file)

Returns a DocAudit object that is embedded in the graph's meta section.
"""

from __future__ import annotations
import os
from datetime import datetime
from graph.types import FileNode, DocAudit, DocWarning

# Categories that count as 'source' for doc-check purposes
_SOURCE_CATEGORIES = {"python", "javascript", "typescript", "react", "shell"}
# Extensions excluded from the 'source file' count inside a directory
_IGNORED_EXTS = {".md", ".mdx", ".txt", ".lock", ".log", ".png", ".jpg", ".ico"}


def _mtime_ts(iso_str: str | None) -> float | None:
    if not iso_str:
        return None
    try:
        return datetime.fromisoformat(iso_str.replace("Z", "+00:00")).timestamp()
    except Exception:
        return None


def audit_docs(root_dir: str, nodes: list[FileNode]) -> DocAudit:
    """
    Run all documentation health checks.

    root_dir  -- project root (used only for display; not re-walked)
    nodes     -- already-parsed FileNode list from project_parser
    """
    warnings: list[DocWarning] = []

    # Build per-directory view: dir_rel → {readme_mtime, readme_path, source_files}
    dir_map: dict[str, dict] = {}

    for node in nodes:
        dir_rel = os.path.dirname(node.path).replace("\\", "/") or "."

        if dir_rel not in dir_map:
            dir_map[dir_rel] = {
                "readme_mtime": None,
                "readme_path": None,
                "source_files": [],
            }

        entry = dir_map[dir_rel]
        filename = os.path.basename(node.path).lower()

        if filename == "readme.md":
            entry["readme_mtime"] = _mtime_ts(node.modified)
            entry["readme_path"] = node.path
        elif node.ext not in _IGNORED_EXTS:
            entry["source_files"].append({
                "id": node.id,
                "path": node.path,
                "mtime": _mtime_ts(node.modified),
            })

    # ── Check 1: Missing README ───────────────────────────────────────────────
    for dir_rel, entry in dir_map.items():
        if not entry["source_files"]:
            continue   # dir has no source — skip
        if entry["readme_path"] is None:
            warnings.append(DocWarning(
                type="missing-readme",
                severity="warning",
                dir=dir_rel,
                message=f"No README.md in '{dir_rel}'",
                affected_files=[f["id"] for f in entry["source_files"]],
            ))
            continue

        # ── Check 2: Stale README ─────────────────────────────────────────────
        stale = [
            f for f in entry["source_files"]
            if f["mtime"] is not None and f["mtime"] > entry["readme_mtime"]
        ]
        if stale:
            latest_ts = max(f["mtime"] for f in stale)
            warnings.append(DocWarning(
                type="stale-readme",
                severity="info",
                dir=dir_rel,
                readme_path=entry["readme_path"],
                message=f"README.md in '{dir_rel}' is older than {len(stale)} file(s)",
                affected_files=[f["id"] for f in stale],
                stale_since=datetime.fromtimestamp(latest_ts).isoformat(),
            ))

    # ── Check 3: Empty module ─────────────────────────────────────────────────
    for node in nodes:
        if node.category not in _SOURCE_CATEGORIES:
            continue
        no_imports = len(node.imports) == 0
        no_exports = len(node.exports) == 0
        no_defs    = len(node.definitions) == 0
        if no_imports and no_exports and no_defs:
            warnings.append(DocWarning(
                type="empty-module",
                severity="info",
                node_id=node.id,
                message=f"'{node.path}' has no imports, exports, or definitions",
                affected_files=[node.id],
            ))

    missing  = sum(1 for w in warnings if w.type == "missing-readme")
    stale    = sum(1 for w in warnings if w.type == "stale-readme")
    empty    = sum(1 for w in warnings if w.type == "empty-module")

    return DocAudit(
        healthy=len(warnings) == 0,
        missing_readmes=missing,
        stale_readmes=stale,
        empty_modules=empty,
        total_warnings=len(warnings),
        warnings=warnings,
    )

# ── GPLv3 interactive notice ──────────────────────────────────────────────────

_GPLv3_WARRANTY = (
    "THERE IS NO WARRANTY FOR THE PROGRAM, TO THE EXTENT PERMITTED BY\n"
    "APPLICABLE LAW. EXCEPT WHEN OTHERWISE STATED IN WRITING THE COPYRIGHT\n"
    'HOLDERS AND/OR OTHER PARTIES PROVIDE THE PROGRAM \"AS IS\" WITHOUT\n'
    "WARRANTY OF ANY KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING, BUT NOT\n"
    "LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A\n"
    "PARTICULAR PURPOSE. THE ENTIRE RISK AS TO THE QUALITY AND PERFORMANCE\n"
    "OF THE PROGRAM IS WITH YOU.  (GPL-3.0-or-later §15)"
)

_GPLv3_CONDITIONS = (
    "You may convey verbatim copies of the Program's source code as you\n"
    "receive it, in any medium, provided that you conspicuously and\n"
    "appropriately publish on each copy an appropriate copyright notice and\n"
    "disclaimer of warranty. (See GPL-3.0 §4-6 for full conditions.)\n"
    "Full license: <https://www.gnu.org/licenses/gpl-3.0.html>"
)


def gplv3_notice():
    """Print the short GPLv3 startup notice. Call this at program startup."""
    print("S-IDE  Copyright (C) 2026  N0V4-N3XU5")
    print("This program comes with ABSOLUTELY NO WARRANTY; for details type 'show w'.")
    print("This is free software, and you are welcome to redistribute it")
    print("under certain conditions; type 'show c' for details.")


def gplv3_handle(cmd: str) -> bool:
    """
    Check whether *cmd* is a GPLv3 license command and handle it.
    Returns True if the command was consumed (caller should skip normal processing).
    """
    match cmd.strip().lower():
        case "show w":
            print(_GPLv3_WARRANTY)
            return True
        case "show c":
            print(_GPLv3_CONDITIONS)
            return True
    return False

# ─────────────────────────────────────────────────────────────────────────────
